src/thunderbots/software/geom/util.cpp

```diff
@@ -0,0 +1,965 @@
+    }
+    return std::fabs((b - a).cross(c - a)) < EPS;
+}
+
+Vector clipPoint(const Vector &p, const Vector &bound1, const Vector &bound2)
+{
+    const double minx = std::min(bound1.x(), bound2.x());
+    const double miny = std::min(bound1.y(), bound2.y());
+    const double maxx = std::max(bound1.x(), bound2.x());
+    const double maxy = std::max(bound1.y(), bound2.y());
+    Vector ret        = p;
+    if (p.x() < minx)
+    {
+        ret.set(minx, ret.y());
+    }
+    else if (p.x() > maxx)
+    {
+        ret.set(maxx, ret.y());
+    }
+    if (p.y() < miny)
+    {
+        ret.set(ret.x(), miny);
+    }
+    else if (p.y() > maxy)
+    {
+        ret.set(ret.x(), maxy);
+    }
+    return ret;
+}
+
+Vector clipPoint(const Vector &p, const Rect &r)
+{
+    const double minx = r.swCorner().x();
+    const double miny = r.swCorner().y();
+    const double maxx = r.neCorner().x();
+    const double maxy = r.neCorner().y();
+    Vector ret        = p;
+    if (p.x() < minx)
+    {
+        ret.set(minx, ret.y());
+    }
+    else if (p.x() > maxx)
+    {
+        ret.set(maxx, ret.y());
+    }
+    if (p.y() < miny)
+    {
+        ret.set(ret.x(), miny);
+    }
+    else if (p.y() > maxy)
+    {
+        ret.set(ret.x(), maxy);
+    }
+    return ret;
+}
+
+// TODO: does this work? unit test
+std::vector<Vector> lineCircleIntersect(
+    const Vector &centre, double radius, const Vector &segA, const Vector &segB)
+{
+    std::vector<Vector> ans;
+
+    // take care of 0 length segments too much error here
+    if ((segB - segA).lensq() < EPS)
+    {
+        return ans;
+    }
+
+    double lenseg = (segB - segA).dot(centre - segA) / (segB - segA).len();
+    Vector C      = segA + lenseg * (segB - segA).norm();
+
+    // if C outside circle no intersections
+    if ((C - centre).lensq() > radius * radius + EPS)
+    {
+        return ans;
+    }
+
+    // if C on circle perimeter return the only intersection
+    if ((C - centre).lensq() < radius * radius + EPS &&
+        (C - centre).lensq() > radius * radius - EPS)
+    {
+        ans.push_back(C);
+        return ans;
+    }
+    // first possible intersection
+    double lensegb = radius * radius - (C - centre).lensq();
+
+    ans.push_back(C - lensegb * (segB - segA).norm());
+    ans.push_back(C + lensegb * (segB - segA).norm());
+
+    return ans;
+}
+
+std::vector<Vector> lineRectIntersect(
+    const Rect &r, const Vector &segA, const Vector &segB)
+{
+    std::vector<Vector> ans;
+    for (unsigned int i = 0; i < 4; i++)
+    {
+        const Vector &a = r[i];
+        const Vector &b = r[i + 1];
+        if (intersects(Seg(a, b), Seg(segA, segB)) &&
+            uniqueLineIntersects(a, b, segA, segB))
+        {
+            ans.push_back(lineIntersection(a, b, segA, segB));
+        }
+    }
+    return ans;
+}
+
+Vector vectorRectIntersect(const Rect &r, const Vector &vecA, const Vector &vecB)
+{
+    /*std::cout << vecA << vecB << r.neCorner() << r.swCorner();
```

Comment:
Dead code. Unless it's useful and should be uncommented, delete it.

Reply:
bump

Reply:
done